Sized self breaks automock

Using 0.10.1, this does not work:

```rust
#[mockall::automock]
pub trait SizedBreaksMock {
    fn new() -> Self
    where
        Self: Sized;
}
```

The use case for this is that the above function may be part of a larger trait,
as a way to make the trait object safe by opting out of making `new()` dispatchable on a dynamic object of the trait:
https://doc.rust-lang.org/reference/items/traits.html#object-safety

I've only tried the latest version.

This is the error dump:

```
error[E0214]: parenthesized type parameters may only be used with a `Fn` trait
 --> a/tests/mocked.rs:9:1
  |
9 | #[mockall::automock]
  | ^^^^^^^^^^^^^^^^^^^^ only `Fn` traits may use parentheses
  |
  = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0107]: missing generics for associated function `Key::new`
    --> a/tests/mocked.rs:9:1
     |
9    | #[mockall::automock]
     | ^^^^^^^^^^^^^^^^^^^^ expected 1 type argument
     |
note: associated function defined here, with 1 type parameter: `T`
    --> /home/a/.cargo/registry/src/github.com-1ecc6299db9ec823/mockall-0.10.1/src/lib.rs:1535:12
     |
1535 |     pub fn new<T: 'static>() -> Self {
     |            ^^^ -
     = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
help: use angle brackets to add missing type argument
     |
9    | #[mockall::automock]::<T>
     |                     ^^^^^

error[E0308]: mismatched types
 --> a/tests/mocked.rs:9:1
  |
9 | #[mockall::automock]
  | ^^^^^^^^^^^^^^^^^^^^ expected struct `Key`, found fn item
  |
  = note: expected reference `&Key`
             found reference `&fn() -> Key {Key::new::<_>}`
  = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
 --> a/tests/mocked.rs:9:1
  |
9 | #[mockall::automock]
  | ^^^^^^^^^^^^^^^^^^^^ expected struct `Key`, found fn item
  |
  = note: expected struct `Key`
            found fn item `fn() -> Key {Key::new::<_>}`
  = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
help: use parentheses to call this function
  |
9 | #[mockall::automock]()
  |                     ^^

error: aborting due to 4 previous errors
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,12 +3,20 @@
 All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/).
 
+## [ Unreleased ] - ReleaseDate
+
+### Fixed
+
+- Fix mocking specializing methods of non-generic structs, a regression in
+  v0.10.0.
+  ([#309](https://github.com/asomers/mockall/pull/309))
+
 ## [ 0.10.1 ] - 2021-07-01
 
 ### Fixed
 
 - Fix mocking trait methods whose return values have lifetime parameters, a
-  regression since v0.10.0.
+  regression in v0.10.0.
   ([#304](https://github.com/asomers/mockall/pull/304))
 
 ## [ 0.10.0 ] - 2021-06-27

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,7 +5,7 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [ 0.10.1 ] - 2021-07-01
 
-## Fixed
+### Fixed
 
 - Fix mocking trait methods whose return values have lifetime parameters, a
   regression since v0.10.0.

--- a/mockall/tests/automock_specializing_method_of_nongeneric_struct.rs
+++ b/mockall/tests/automock_specializing_method_of_nongeneric_struct.rs
@@ -1,0 +1,29 @@
+// vim: tw=80
+//! Non-generic structs can have specializing methods, too.  For example, some
+//! methods place constraints on Self.  It's even possible for a where clause to
+//! place a constraint on types that appear nowhere in the struct's signatures.
+#![deny(warnings)]
+
+use mockall::*;
+
+#[automock]
+trait Bar {
+    fn bar(&self) where Self: Sized;
+    fn baz() where Self: Sized;
+}
+
+#[test]
+fn nonstatic() {
+    let mut mock = MockBar::default();
+    mock.expect_bar()
+        .return_const(());
+    mock.bar();
+}
+
+#[test]
+fn static_method() {
+    let ctx = MockBar::baz_context();
+    ctx.expect()
+        .return_const(());
+    MockBar::baz();
+}

--- a/mockall_derive/src/lib.rs
+++ b/mockall_derive/src/lib.rs
@@ -753,7 +753,7 @@ fn supersuperfy_bounds(
 /// Generate a suitable mockall::Key generic paramter from any Generics
 fn gen_keyid(g: &Generics) -> impl ToTokens {
     match g.params.len() {
-        0 => quote!(),
+        0 => quote!(<()>),
         1 => {
             let (_, tg, _) = g.split_for_impl();
             quote!(#tg)
@@ -1351,7 +1351,7 @@ mod gen_keyid {
 
     #[test]
     fn empty() {
-        check_gen_keyid(quote!(), quote!());
+        check_gen_keyid(quote!(), quote!(<()>));
     }
 
     #[test]

